File: backend/src/main/java/com/smartproject/platform/service/PermissionService.java
package com.smartproject.platform.service;

import com.smartproject.platform.model.*;
import com.smartproject.platform.repository.PermissionRepository;
import com.smartproject.platform.repository.ProjectMemberRoleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class PermissionService {
    
    private final PermissionRepository permissionRepository;
    private final ProjectMemberRoleRepository projectMemberRoleRepository;
    
    /**
     * Check if user has permission in a project context
     */
    public boolean hasPermission(User user, Project project, String permissionName) {
        // 1. Check global roles first (ADMIN has all permissions)
        if (hasGlobalPermission(user, permissionName)) {
            return true;
        }
        
        // 2. Check project-specific role
        Optional<ProjectMemberRole> projectRole = projectMemberRoleRepository
                .findByProjectIdAndUserId(project.getId(), user.getId());
        
        if (projectRole.isPresent()) {
            return hasRolePermission(projectRole.get().getRole(), permissionName);
        }
        
        return false;
    }
    
    /**
     * Check if user can perform action based on their role hierarchy
     */
    public boolean canPerformAction(User user, Project project, String action) {
        Role userRole = getEffectiveRole(user, project);
        return canRolePerformAction(userRole, action);
    }
    
    /**
     * Get effective role for user in project context
     */
    public Role getEffectiveRole(User user, Project project) {
        // Check global roles first
        if (user.getRoles().contains(Role.ADMIN)) {
            return Role.ADMIN;
        }
        
        // Check project-specific role
        Optional<ProjectMemberRole> projectRole = projectMemberRoleRepository
                .findByProjectIdAndUserId(project.getId(), user.getId());
        
        if (projectRole.isPresent()) {
            return projectRole.get().getRole();
        }
        
        // Default to USER if no project role assigned
        return Role.USER;
    }
    
    /**
     * Check if user has global permission (not project-specific)
     */
    private boolean hasGlobalPermission(User user, String permissionName) {
        if (user.getRoles().contains(Role.ADMIN)) {
            return true; // Admin has all permissions
        }
        
        // Check if any of user's roles have this permission
        for (Role role : user.getRoles()) {
            if (hasRolePermission(role, permissionName)) {
                return true;
            }
        }
        
        return false;
    }
    
    /**
     * Check if a role has a specific permission
     */
    private boolean hasRolePermission(Role role, String permissionName) {
        List<Permission> rolePermissions = permissionRepository.findByRole(role.name());
        return rolePermissions.stream()
                .anyMatch(p -> p.getName().equals(permissionName));
    }
    
    /**
     * Check if role can perform action based on hierarchy
     */
    private boolean canRolePerformAction(Role role, String action) {
        int roleLevel = getRoleHierarchy(role);
        
        return switch (action) {
            case "CREATE_EPIC" -> roleLevel >= getRoleHierarchy(Role.SENIOR);
            case "CREATE_MILESTONE" -> roleLevel >= getRoleHierarchy(Role.SENIOR);
            case "CREATE_TASK" -> roleLevel >= getRoleHierarchy(Role.SDE_2);
            case "ASSIGN_TASK" -> roleLevel >= getRoleHierarchy(Role.SDE_2);
            case "APPROVE_PR" -> roleLevel >= getRoleHierarchy(Role.SENIOR);
            case "APPROVE_DESIGN" -> roleLevel >= getRoleHierarchy(Role.STAFF);
            case "APPROVE_ARCHITECTURE" -> roleLevel >= getRoleHierarchy(Role.ARCHITECT);
            case "ESTIMATE_TIMELINE" -> roleLevel >= getRoleHierarchy(Role.SDE_2);
            case "OVERRIDE_DECISION" -> roleLevel >= getRoleHierarchy(Role.STAFF);
            case "UPDATE_TASK_STATUS" -> roleLevel >= getRoleHierarchy(Role.SDE_1);
            default -> false;
        };
    }
    
    /**
     * Check if user role is at least the minimum required role
     */
    public boolean isAtLeast(User user, Project project, Role minimumRole) {
        Role userRole = getEffectiveRole(user, project);
        return getRoleHierarchy(userRole) >= getRoleHierarchy(minimumRole);
    }
    
    /**
     * Get role hierarchy level
     */
    private int getRoleHierarchy(Role role) {
        return switch (role) {
            case SDE_1 -> 1;
            case SDE_2 -> 2;
            case SENIOR -> 3;
            case STAFF -> 4;
            case PRINCIPAL -> 5;
            case ARCHITECT -> 6;
            case ADMIN -> 10; // Admin has all permissions
            case TECHNICAL_LEAD -> 3; // Map to SENIOR level
            case MANAGER -> 2; // Map to SDE_2 level
            case USER -> 1; // Map to SDE_1 level
        };
    }
    
    /**
     * Check if user can view entity based on visibility level
     */
    public boolean canViewEntity(User user, Project project, VisibilityLevel visibilityLevel) {
        Role userRole = getEffectiveRole(user, project);
        
        return switch (visibilityLevel) {
            case PUBLIC -> true;
            case TEAM -> getRoleHierarchy(userRole) >= getRoleHierarchy(Role.SDE_2);
            case SENIOR_PLUS -> getRoleHierarchy(userRole) >= getRoleHierarchy(Role.SENIOR);
            case STAFF_PLUS -> getRoleHierarchy(userRole) >= getRoleHierarchy(Role.STAFF);
            case PRINCIPAL_PLUS -> getRoleHierarchy(userRole) >= getRoleHierarchy(Role.PRINCIPAL);
        };
    }
}

File: frontend/src/app/core/permission.service.ts
import { Injectable } from '@angular/core';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class PermissionService {
  private roleHierarchy: { [key: string]: number } = {
    'SDE_1': 1,
    'SDE_2': 2,
    'SENIOR': 3,
    'STAFF': 4,
    'PRINCIPAL': 5,
    'ARCHITECT': 6,
    'ADMIN': 10,
    'TECHNICAL_LEAD': 3, // Map to SENIOR level
    'MANAGER': 2, // Map to SDE_2 level
    'USER': 1 // Map to SDE_1 level
  };

  constructor(private authService: AuthService) {}

  /**
   * Get user's role from current user or project context
   */
  getUserRole(projectRole?: string): string {
    if (projectRole) {
      return projectRole;
    }
    const user = this.authService.currentUserValue;
    if (user && user.roles && user.roles.length > 0) {
      // Return highest role
      return user.roles.reduce((highest: string, role: string) => {
        const currentLevel = this.roleHierarchy[role] || 0;
        const highestLevel = this.roleHierarchy[highest] || 0;
        return currentLevel > highestLevel ? role : highest;
      }, user.roles[0]);
    }
    return 'USER';
  }

  /**
   * Check if user can perform action
   */
  canPerformAction(action: string, projectRole?: string): boolean {
    const userRole = this.getUserRole(projectRole);
    const requiredRole = this.getRequiredRole(action);
    const userLevel = this.roleHierarchy[userRole] || 0;
    const requiredLevel = this.roleHierarchy[requiredRole] || 0;
    return userLevel >= requiredLevel;
  }

  /**
   * Get required role for an action
   */
  private getRequiredRole(action: string): string {
    const actionMap: { [key: string]: string } = {
      'CREATE_EPIC': 'SENIOR',
      'CREATE_MILESTONE': 'SENIOR',
      'CREATE_TASK': 'SDE_2',
      'ASSIGN_TASK': 'SDE_2',
      'APPROVE_PR': 'SENIOR',
      'APPROVE_DESIGN': 'STAFF',
      'APPROVE_ARCHITECTURE': 'ARCHITECT',
      'ESTIMATE_TIMELINE': 'SDE_2',
      'OVERRIDE_DECISION': 'STAFF',
      'UPDATE_TASK_STATUS': 'SDE_1',
      'REQUEST_APPROVAL': 'SDE_1',
      'APPROVE_APPROVALS': 'SENIOR',
      'VIEW_APPROVALS': 'SDE_1',
      'VIEW_DECISIONS': 'SDE_1'
    };
    return actionMap[action] || 'ARCHITECT';
  }

  /**
   * Check if user can create epics
   */
  canCreateEpic(projectRole?: string): boolean {
    return this.canPerformAction('CREATE_EPIC', projectRole);
  }

  /**
   * Check if user can create milestones
   */
  canCreateMilestone(projectRole?: string): boolean {
    return this.canPerformAction('CREATE_MILESTONE', projectRole);
  }

  /**
   * Check if user can assign tasks
   */
  canAssignTasks(projectRole?: string): boolean {
    return this.canPerformAction('ASSIGN_TASK', projectRole);
  }

  /**
   * Check if user can approve PRs
   */
  canApprovePR(projectRole?: string): boolean {
    return this.canPerformAction('APPROVE_PR', projectRole);
  }

  /**
   * Check if user can approve architecture decisions
   */
  canApproveArchitecture(projectRole?: string): boolean {
    return this.canPerformAction('APPROVE_ARCHITECTURE', projectRole);
  }

  /**
   * Check if user can request approval
   */
  canRequestApproval(projectRole?: string): boolean {
    return this.canPerformAction('REQUEST_APPROVAL', projectRole);
  }

  /**
   * Check if user can approve approvals
   */
  canApproveApprovals(projectRole?: string): boolean {
    return this.canPerformAction('APPROVE_APPROVALS', projectRole);
  }

  /**
   * Check if user can view approvals queue
   */
  canViewApprovals(projectRole?: string): boolean {
    return this.canPerformAction('VIEW_APPROVALS', projectRole);
  }

  /**
   * Check if user can view decisions
   */
  canViewDecisions(projectRole?: string): boolean {
    return this.canPerformAction('VIEW_DECISIONS', projectRole);
  }

  /**
   * Check if role is at least minimum required
   */
  isAtLeast(userRole: string, minimumRole: string): boolean {
    const userLevel = this.roleHierarchy[userRole] || 0;
    const minLevel = this.roleHierarchy[minimumRole] || 0;
    return userLevel >= minLevel;
  }

  /**
   * Get role display name
   */
  getRoleDisplayName(role: string): string {
    const displayNames: { [key: string]: string } = {
      'SDE_1': 'SDE-1',
      'SDE_2': 'SDE-2',
      'SENIOR': 'Senior Engineer',
      'STAFF': 'Staff Engineer',
      'PRINCIPAL': 'Principal Engineer',
      'ARCHITECT': 'Software Architect'
    };
    return displayNames[role] || role;
  }
}
